roaring/src/bitmap/iter.rs

```diff
@@ -294,7 +316,306 @@ impl RoaringBitmap {
+        fn iter_back_mut(&mut self) -> &mut Option<container::Iter<'a>>;
+
+        fn dec_size_hint(&mut self);
+
+        fn empty_inner_iter() -> container::Iter<'static>;
+
+        fn decompose(
+            self,
+        ) -> (container::Iter<'a>, Option<container::Iter<'a>>, Self::IntoContainerIterator);
+
+        #[inline]
+        fn next_inner(&mut self) -> Option<u32> {
+            self.dec_size_hint();
+            loop {
+                if let item @ Some(_) = self.iter_front_mut().next() {
+                    return item;
+                }
+                if self.advance_container().is_some() {
+                    continue;
+                } else {
+                    return None;
+                }
+            }
+        }
+
+        #[inline]
+        fn next_back_inner(&mut self) -> Option<u32> {
+            self.dec_size_hint();
+            loop {
+                if let Some(iter) = self.iter_back_mut() {
+                    if let item @ Some(_) = iter.next_back() {
+                        return item;
+                    }
+                }
+                if self.advance_container_back().is_some() {
+                    continue;
+                } else {
+                    return None;
+                }
+            }
+        }
+
+        fn advance_to_inner(&mut self, n: u32) {
+            let (key, _) = util::split(n);
+            fn advance_iter(iter: &mut container::Iter<'_>, n: u32) -> Option<u32> {
+                while let Some(next) = iter.peek() {
+                    if next < n {
+                        iter.next();
+                    } else {
+                        return Some(next);
+                    }
+                }
+                None
+            }
+            if let Some(index) = self.find_container(key) {
+                self.drain_containers_until(index);
+                let container = self.pop_container_front().expect("bug!");
+                let mut iter = container.into_iter();
+                advance_iter(&mut iter, n);
+                *self.iter_front_mut() = iter;
+            } else {
+                // there are no containers with given key. Look in iter_front and iter_back.
+                let iter_front = self.iter_front_mut();
+                if iter_front.peek().map(|n| util::split(n).0) == Some(key) {
+                    advance_iter(iter_front, n);
+                    return;
+                }
+                if let Some(iter_back) = self.iter_back_mut() {
+                    if iter_back.peek().map(|n| util::split(n).0) == Some(key) {
+                        advance_iter(iter_back, n);
+                        *self.iter_front_mut() = Self::empty_inner_iter();
+                        return;
+                    }
+                }
+                self.clear_containers();
+                *self.iter_front_mut() = Self::empty_inner_iter();
+                *self.iter_back_mut() = None;
+            }
+        }
```

Comment:
I'm pretty sure this isn't updating the `size_hint` for the iterator.

I really think trying to keep track of the `size_hint` for these iterators was a mistake, and would rather just remove the exact size_hint tracking.

Reply:
You're completely right of course. I've fixed this in the latest commit. But I agree, maybe it would be better to remove this code. It certainly would make it cleaner.